imap_processing/spice/geometry.py

```diff
@@ -159,3 +161,102 @@ def get_spacecraft_spin_phase(
     # now (dummy implementation)
     # TODO: Replace with actual logic to calculate spin phase
     return np.array(query_met_times, dtype=float)
+
+
+@typing.no_type_check
+@ensure_spice
+def frame_transform(
+    from_frame: SpiceFrame,
+    to_frame: SpiceFrame,
+    et: Union[float, npt.NDArray],
+    position: npt.NDArray,
+) -> npt.NDArray:
+    """
+    Transform an <x, y, z> vector between reference frames (rotation only).
+
+    Parameters
+    ----------
+    from_frame : SpiceFrame
+        Reference frame of input vector(s).
+    to_frame : SpiceFrame
+        Reference frame of output vector(s).
+    et : float or npt.NDArray
+        Ephemeris time(s) corresponding to position(s).
+    position : npt.NDArray
+        <x, y, z> vector or array of vectors in reference frame `from_frame`.
+
+    Returns
+    -------
+    result : npt.NDArray
+        3d position vector(s) in reference frame `to_frame`.
+    """
+    if position.ndim == 1:
+        if not len(position) == 3:
+            raise ValueError(
+                "Position vectors with one dimension must have 3 elements."
+            )
+        if not isinstance(et, float):
+            raise ValueError(
+                "Ephemeris time must be float when single position vector is provided."
+            )
+    elif position.ndim == 2:
+        if not position.shape[1] == 3:
+            raise ValueError(
+                f"Invalid position shape: {position.shape}. "
+                f"Each input position vector must have 3 elements."
+            )
+        if not len(position) == len(et):
+            raise ValueError(
+                "Mismatch in number of position vectors and Ephemeris times provided."
+                f"Position has {len(position)} elements and et has {len(et)} elements."
+            )
+
+    rotate = get_rotation_matrix(from_frame, to_frame, et)
+
+    if hasattr(rotate[0][0], "__len__"):
+        result = np.array(
+            [
+                np.matmul(rotate, pos).astype(np.float64)
+                for rotate, pos in zip(rotate, position)
+            ]
+        )
```

Comment:
Could you add a comment about what the shapes of the arrays you're expecting in these two cases are?

I think this is the array path and that you could broadcast the matmul operations, expanding/shrinking extra dimensions as needed to avoid the for loop here.

```python
# (i, n, k) @ (i, k, m) -> (i, n, m)
np.array(rotate, dtype=np.float64) @ np.array(position, dtype=np.float64)
```

Reply:
Yes! Thanks for commenting on this. I had identified this as something to investigate when I copied the code but forgot to do that. I also learned about the `@` operator!